### DISCUSSION TITLE
V3 Simulation Boundary Condition Influence

### DISCUSSION THREAD

**Discussion by nrz22** [NatLabRockies/floris#912] [Q&A] [answered]
Hello, 

I am quite new to Python and FLORIS, but I am using it for my master's thesis to investigate the effects of wind farm clustering. 
What I have done so far, is to create a wind farm layout generator, such that I can build two wind farms and superimpose them into one array to feed into FLORIS. 

I have been modifying EX 29 in V3 to calculate power and visualize the wakes for different farm layouts and turbine sizes. 
I have gotten varying power results that in my head are different in my head than I am expecting for different scenarios, such as an increase in power when the farms are closer together vs. when they are further apart, etc. 

I am wondering, as I know the way the boundary conditions are treated in the calculations plays a role in the output, how the boundary is treated when using fi.calculate_wake() and fi.get_turbine_powers(). 

I am currently running these simulations with the emgauss_floating.yaml that I updated to use either the IEA_10MW or 15MW turbine. 

I have read through the documentation on tools/floris_interface.py and simulation/flow_field.py to try and understand better, but my understanding of python is not great enough to fully interpret what is happening. 

Thank you for your time.

**Comment by christiannvaughn** [comment]
Hi @nrz22, our team is out of office this week, but we will try to get back to you asap.

**Comment by misi9170** [comment, marked as answer]
Hi @nrz22, thanks for your question. Generally speaking, I agree that you will usually see higher power when turbines are spaced further apart (in the direction of the flow).

I'm not 100% sure what you mean by boundary conditions. FLORIS does not solve for the flow in a mesh, as CFD solvers do---rather, it computes analytically the thrust produced by turbines and their corresponding wakes, and evaluates those wakes only at the turbine locations so that power may be evaluated. For creating flow visualizations, we do sample the wakes across a grid of points, but still we do not solve any Navier-Stokes equations or any other sort of differential equations across the grid that would require a boundary condition.

Rather, the free stream wind speed is used to calculate the inflow to any unwaked turbines; and the (analytical) velocities in the wakes are used to calculate the inflow to any waked turbines (turbines operating in the wake, or partially in the wake, of other turbines for a given wind direction).

I hope this helps. If it's not clear, I'm happy to try to explain further. It would be helpful if you could provide an example of a script that you are running to demonstrate the unexpected behavior.

**Comment by nrz22** [comment]
@misi9170 Hello, thank you for your response. 

Your explanation does make sense, however, I still have a few questions.
First, can you guide me to where the turbine thrust and velocity calculations take place in the FLORIS code so I can study and understand what is happening, please. 

And then if I am correct, then unlike CFD there can be no wind tunnel effect in the power calculations because it is solving steady-state and at the individual turbines, so the size of the flow field around the wind farm should not affect the calculations? And then if that's true, I am wondering how the velocity is calculated downstream each turbine and when it eventually reaches another turbine? That will depend on the velocity deficit and combination models being used I believe, but I am also curious where I can see that implemented to wrap my head around it. 

In regards to trying to use FLORIS to do power calculations between two wind farms, I have been introducing the two farms to FLORIS as in V3 example 09 where it is input as a single array. I believe this is the only way to do it, but I'm curious if it is not and if there is a best practice for doing this in order to achieve some results that make some sense, because right now I am seeing multiple GW differences in stand-alone farms and clustered ones depending on distance and organization that are a little backward of what I'm expecting (and my expectations can easily be wrong). 

For my code I have attached the modified EX 29 I have been using to get power and do some visualizations. I am also using some functions I created to build my windfarms but below I attached the output arrays of running the EX file with 15MW turbines, alignment north and south of the turbines in each farm, and the two farms as close as they can be to each other, which is roughly 14km.
I also included some power data from increasing the distances. When they are aligned like this the change is not as much as in some cases I have modeled. 

Thank you for your time and help. Cheers. 

[Test_Visual_Power_MOD29 - txt.txt](https://github.com/user-attachments/files/15546288/Test_Visual_Power_MOD29.-.txt.txt)
[Cluster Power results at varying distances.txt](https://github.com/user-attachments/files/15547800/Cluster.Power.results.at.varying.distances.txt)

**Reply by misi9170** [reply]
Hi @nrz22 ,

1. In Floris v3, the thrust is evaluated in the main solver code [here](https://github.com/NREL/floris/blob/v3/floris/simulation/solver.py#L104), which calls the thrust function implemented [here](https://github.com/NREL/floris/blob/v3/floris/simulation/turbine.py#L221). Power is not evaluated during the main solve, but only after the solve [here](https://github.com/NREL/floris/blob/v3/floris/tools/floris_interface.py#L613), which calls the power function [here](https://github.com/NREL/floris/blob/v3/floris/simulation/turbine.py#L168). The process is conceptually the same in Floris v4, but the structure of the package has changed.

2. That's correct---the size of the flow field does not affect the calculations, and in fact there is generally no "flow field" as such---the flow is only evaluated at the turbine locations. You can see how this works by looking at the main solve routine [here](https://github.com/NREL/floris/blob/v3/floris/simulation/solver.py#L63). Note that this is the solver for the GCH model and a few other models---I notice you are using the Empirical Gaussian (EmG) model in the scripts that you sent---you can find the solver for the EmG model [here](https://github.com/NREL/floris/blob/v3/floris/simulation/solver.py#L1154).

3. Thanks for sending through the scripts and outputs. I'm not able to run the script without the extra data (the `test_wind_farm` package that you are importing), but I see that there are two farms of 33 turbines each. If I'm understanding correctly, then one farm is directly north of the other. Do the two wind farms have identical layouts? Could you share an image of those layouts perhaps? I'm assuming that the 14km refers to the distance between the two wind farms (rather than the spacing between turbines in a single wind farm). 14km is approximately 60 rotor diameters for the IEA 15MW turbine. This is quite a long distance, so the wake will have almost completely recovered if you are using the default settings for the EmG model. That may explain why you are not seeing much change in the powers as the spacing grows. You could consider trying much closer spacings (say, 1km and 2km) and see if you see the expected trend between those two. You could also consider adding a slower recovery zone to the EmG model input file, as demonstrated [here](https://github.com/NREL/floris/blob/v3/examples/26_empirical_gauss_velocity_deficit_parameters.py#L172), to see the wakes persist longer.

**Reply by nrz22** [reply]
Thank you for your response. That helps clear up my confusion. 

I have lately been trying to understand how the model parameters affect the outcome of the simulations, and I have been focusing on the GCH model. I am wondering if you can explain to me where the alpha and beta constants come from. I understand that they are values for determining where the far wake begins, and I swear I read very recently a paper that had these values explained in it, but I cannot find any such paper at this moment. 

I am curious about these values though, mainly the beta, as in the gaussian model description, beta is defined as function of the thrust coefficient, and I would like to know why it was adopted it to be a constant in this model and what the significance of that is on the results, or if it is an entirely different variable beta. 

Your help is very much appreciated. Thank you again for your quick responses and insightful answers. Cheers.

**Reply by nrz22** [reply]
I am also curious about the Crespo-Hernandez tuning variables. I know there is an equation for the relationship, and that you guys have four variables defined for the Initial, constant, axial induction, and downstream. 

The part I am curious about is why the constant value is different in the GCH, empgauss, Jensen, Turbopark at 0.5 and is 0.9 for the CC model.

**Reply by misi9170** [reply]
@nrz22 , alpha and beta are from [Bastankhah and Porté-Agel's 2016 paper](https://www.cambridge.org/core/journals/journal-of-fluid-mechanics/article/experimental-and-theoretical-study-of-wind-turbine-wakes-in-yawed-conditions/400859134E38F340C8811FD7AAD0CE05), see in particular equation (6.16).

Regarding the Crespo-Hernandez model tuning for the CC model, you could look at Section 3.2.1 of [Bay et al. (2023)](https://wes.copernicus.org/articles/8/401/2023/wes-8-401-2023.pdf); there aren't really details there of why exactly the constant of 0.9 was chosen, but I'd guess that this was needed to best match the CC model results with SOWFA data. @bayc, do you have any further insights here?

**Reply by nrz22** [reply]
Hello, thank you for your response, it has been very helpful. 

I am wondering now, how reliable my study can be using the default parameters within FLORIS, as I do not have any real data for my location to do any matching. 
I was curious how much of an effect each of these parameters would have on the power output for a simulation, I have done a small sensitivity analysis changing each variable independently by +/- 25% for the gauss and crespo hernandez parameters, and have actually found some pretty significant variance in power output of up to 15% (relative difference) for the ka gauss parameter and up to 20% using the constant of 0.9 that the CC model uses for C_H compared to the 0.5 of others, for a row of three 10 MW IEA turbines, at 8D spacing. 

I am pretty new to all of this, and from my reading on others using FLORIS I have just assumed they have been using the default parameters unless otherwise stated, so if you have any recommendations on how to proceed that would be much appreciated. 

Cheers,
Nicholas

**Reply by misi9170** [reply]
Hi Nicholas,

If you don't have any data to compare to, I think you can use the default parameters as a best guess, but of course you should still treat the results carefully. This is discussed further in #757 , and in particular see [this comment](https://github.com/NREL/floris/discussions/757#discussioncomment-7782242). 

You could also take a look at [Doekemeijer et al. (2022)](https://www.mdpi.com/1996-1073/15/6/1964) to get a feel for how different models perform in their default configurations. While the authors do not perform a full sensitivity study of the Crespo-Hernandez wake added turbulence model, the impact of atmospheric turbulence on the models is discussed. 

Thanks,
Misha
